src/app/reporting/Engine.cpp

```diff
@@ -428,6 +430,8 @@ CHIP_ERROR Engine::BuildSingleReportDataAttributeReportIBs(ReportDataMessage::Bu
             // Successfully encoded the attribute, clear the internal state.
             apReadHandler->SetAttributeEncodeState(AttributeEncodeState());
         }
+        iterator.MarkCompleted();
```

Comment:
Shouldn't this happen in the various Continue() calls above too?

It might be better to keep the for loop structure where MarkCompleted() happens as the "advance to next" part of the for loop header (the third part)....  At least in terms of being sure we are preserving behavior.

Reply:
It happens automatically whenever Next() is called again. This is a final mark (and would happen also if I called next() again).

This is awkward, however if I change it to not use PeekIterator and make it just reset the state on every loop end (which is more readable as an intent) it gets larger code size (by a tiny amount, like 20 bytes ... but still larger).

I am fighting hard here between readability vs code compactness preserving.

Reply:
Renamed it to `Rollback` instead of `Peek` and made the `MarkCompleted call mandatory (and moved it within the loop as a for loop advancement). 

This still feels `tricky` however at least Next has less sideffects.